src/AGCertificateVerifier.cpp

```diff
@@ -0,0 +1,973 @@
+
+    // Read CRLSet header (JSON)
+    uint16_t json_len = *(uint16_t *)(pos);
+    pos += 2; // header length
+    jsmntok_t tokens[JSON_MAX_TOKENS];
+    jsmn_parser parser = {0};
+    jsmn_init(&parser);
+    const char *json = (const char *) pos;
+    // JSMN parser is very fast and does not build a document tree - just streamed offsets-only representation
+    int tokens_num = jsmn_parse(&parser, json, json_len, tokens, JSON_MAX_TOKENS);
+    if (tokens_num >= 0) {
+        for (int i = 0; i < tokens_num; i++) {
+            jsmntok_t *key_tok = &tokens[i];
+            // If token is object property key
+            if (key_tok->type == JSMN_STRING && key_tok->size == 1) {
+                const std::string key = std::string(json + key_tok->start, (size_t) key_tok->end - key_tok->start);
+                if (key == BLOCKED_SPKIS_KEY) {
+                    // Found BlockedSPKIs
+                    // Value is in next token
+                    jsmntok_t *value_tok = key_tok + 1;
+                    if (value_tok->type == JSMN_ARRAY) {
+                        // Next `value_tok->size' tokens are array elements
+                        jsmntok_t *t = value_tok + 1;
+                        jsmntok_t *endpos_tok = t + value_tok->size;
+                        while (t != endpos_tok) {
+                            if (t->type == JSMN_STRING) {
+                                const std::string spki = std::string(json + t->start, (size_t) t->end - t->start);
+                                revokedSHADigests.insert(spki);
+                            } else {
+                                break;
+                            }
+                            ++t;
+                        }
+                    }
+                    break;
+                }
+            }
+        }
+    }
+    pos += json_len;
+
+    // Read CRLSet
+    unsigned char *end = (unsigned char *) (crlSetContent.c_str() + len);
+#define BCHECK(x) do { if (!(x)) break; } while (0)
+    while (pos < end) {
+        // Issuer SPKI SHA256 - 32 bytes
+        BCHECK(pos + 32 > end);
+        std::string issuerSHADigest = AGStringUtils::encodeToBase64(pos, 32);
+        pos += 32;
+        BCHECK(pos >= end);
+        // Serial number count - 4 bytes
+        uint32_t serialCount = *(uint32_t *)(pos);
+        pos += 4;
+        BCHECK(pos >= end);
+        // Serial number list
+        std::set<std::string> &serialList = issuerSHADigestToCRL[issuerSHADigest];
+        for (int i = 0; i < serialCount; i++) {
+            // Serial number size - 1 byte
+            uint8_t serialSize = *pos++;
+            BCHECK(pos >= end);
+            BCHECK(pos + serialSize > end);
+            // Serial number
+            serialList.insert(AGStringUtils::encodeToHex(pos, serialSize));
+            pos += serialSize;
+            BCHECK(pos >= end);
+        }
+    }
+#undef BCHECK
+}
+
+/**
+ * Update HPKP pinning information
```

Comment:
There should be a link to HPKP definition.
It's not clear when this method should be called by the client.